lib/autohttp/agent.js

```diff
@@ -0,0 +1,287 @@
+    this.defaultPort = 443
+  }
+
+  createConnection (
+    req,
+    reqOptions,
+    cb,
+    socketCb
+  ) {
+    let options = Object.assign({}, reqOptions, this.options)
+    options = Object.assign(options, {
+      port: Number(options.port || this.defaultPort),
+      host: options.hostname || options.host || 'localhost'
+    })
+
+    // check if ALPN is cached
+    const name = this.getName(options)
+    const [protocol, cachedSocket] = this.ALPNCache.get(name) || []
+
+    if (!protocol || !cachedSocket || cachedSocket.closed || cachedSocket.destroyed) {
+      // No cache exists or the initial socket used to establish the connection has been closed. Perform ALPN again.
+      this.ALPNCache.delete(name)
+      this.createNewSocketConnection(req, options, cb, socketCb)
+      return
+    }
+
+    // No need to pass the cachedSocket since the respective protocol's agents will reuse the socket that was initially
+    // passed during ALPN Negotiation
+    if (protocol === 'h2') {
+      const http2Options = Object.assign({}, options, {
+        path: options.socketPath
+      })
+
+      let connection
+      try {
+        const uri = options.uri
+        connection = this.http2Agent.createConnection(req, uri, http2Options)
+      } catch (e) {
+        cb(e)
+        connection && connection.socket && socketCb(connection.socket)
+        return
+      }
+
+      cb(null, 'http2', connection)
+      socketCb(connection.socket)
+
+      return
+    }
+
+    if (protocol === 'http/1.1' || protocol === 'http/1.0') {
+      const http1RequestOptions = Object.assign({}, options, {
+        agent: this.httpsAgent
+      })
+
+      let request
+      try {
+        request = https.request(http1RequestOptions)
+      } catch (e) {
+        cb(e)
+        return
+      }
+
+      request.on('socket', (socket) => socketCb(socket))
+      cb(null, 'http1', request)
+    }
+  }
+
+  createNewSocketConnection (req, options, cb, socketCb) {
+    const uri = options.uri
+    const name = this.getName(options)
+
+    const tlsSocketOptions = Object.assign({}, options, {
+      path: options.socketPath,
+      ALPNProtocols: ['h2', 'http/1.1', 'http/1.0'],
+      servername: options.servername || calculateServerName(options)
+    })
+
+    const socket = tls.connect(tlsSocketOptions)
+    socketCb(socket)
+
+    const socketConnectionErrorHandler = (e) => {
+      cb(e)
+    }
+    socket.on('error', socketConnectionErrorHandler)
+
+    socket.once('secureConnect', () => {
+      socket.removeListener('error', socketConnectionErrorHandler)
+
+      const protocol = socket.alpnProtocol
+
+      if (!protocol) {
+        cb(socket.authorizationError)
+        socket.end()
+        return
+      }
+
+      if (protocol !== 'h2' && protocol !== 'http/1.1') {
+        cb(new Error('Unknown protocol' + protocol))
+        return
+      }
+
+      // Update the cache
+      this.ALPNCache.set(name, [protocol, socket])
+
+      socket.on('close', () => {
```

Comment:
Can we check this doesn't result in a memory leak?
```suggestion
      socket.once('close', () => {
```

Reply:
Ran profiling. Memory consumption remains constant